[BUGFIX] Fix Postgres, Trino quoted table identifier issues

Kilo59 opened the pull request:
Fix the quoted identifier issue introduced by previous bugfix PRs.
Specifically, this prevents double quoting table names if quotes are part of the original string.

- https://github.com/great-expectations/great_expectations/pull/8327
- https://github.com/great-expectations/great_expectations/pull/8361

https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.quoted_name


------


- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more details, see our [Contribution Checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist), [Coding style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style), and [Documentation style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/docs_style).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

ghost commented:
<details open>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=8442&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

####

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/8442/9eaa7415/482a1e497490b78102b06d1241c22b21bf81e127.svg)](https://app.codesee.io/r/reviews?pr=8442&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
<img src="https://app.codesee.io/Legend_v3.png" alt="CodeSee Map legend" width="220" height="156">
</details>

Kilo59 committed:
check if string has quote strings

" " -> "_" TODO

spel better

Kilo59 committed:
update comment

Kilo59 committed:
`SnowflakeDatasource._default_table_name()` pydantic validator method

Kilo59 commented on great_expectations/datasource/fluent/snowflake_datasource.py:
```diff
@@ -100,6 +100,24 @@ class SnowflakeDatasource(SQLDatasource):
         "warehouse",
     }
 
+    @pydantic.validator("table_name", pre=True, always=True)
+    def _default_table_name(cls, table_name: str, values: dict, **kwargs) -> str:
+        if not (validated_table_name := table_name or values.get("name")):
+            raise ValueError(
+                "table_name cannot be empty and should default to name if not provided"
+            )
+        if cls._is_bracketed_by_quotes(validated_table_name):
+            from great_expectations.compatibility import sqlalchemy
+
+            if sqlalchemy.quoted_name:
+                return sqlalchemy.quoted_name(value=validated_table_name, quote=True)
+        else:
+            # snowflake sql uses uppercase idenetiers, so we uppercase the table name
+            # sqlalchemy will auto convert them but this is more explicit
+            return validated_table_name.upper()
+
+        return validated_table_name
```
These changes to the `SnowflakeDatasource` don't appear to be necessary.

Kilo59 commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -819,13 +819,15 @@ def qualified_name(self) -> str:
         from great_expectations.compatibility import sqlalchemy
 
-        if sqlalchemy.quoted_name:
+        if cls._is_bracketed_by_quotes(validated_table_name) and sqlalchemy.quoted_name:
             return sqlalchemy.quoted_name(value=validated_table_name, quote=True)
```
```suggestion
        if sqlalchemy.quoted_name:
            return sqlalchemy.quoted_name(
                value=validated_table_name,
                quote=cls._is_bracketed_by_quotes(validated_table_name)
               )
```
🤔 need to test this, I think it should work too.

Kilo59 committed:
revert snowflake specific `_default_table_name` validator

Kilo59 committed:
replace user quotes with `sqlalchemy.quoted_name`

Kilo59 commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -819,14 +819,26 @@ def qualified_name(self) -> str:
 
         if sqlalchemy.quoted_name:
-            return sqlalchemy.quoted_name(value=validated_table_name, quote=True)
+            if table_name_is_qouted:
+                # Remove the quotes and add them back using the sqlalchemy.quoted_name function
+                # TODO: do we need to worry about nested qoutes?
+                validated_table_name = validated_table_name.strip("'").strip('"')
+
+            return sqlalchemy.quoted_name(
+                value=validated_table_name,
+                quote=table_name_is_qouted,
+            )
```
This feels a little silly.
I don't totally understand what problem the original `sqlalchemy.quoted_name` was solving.
But I believe this solution preserves the intended fix this was meant to apply.

Kilo59 replied:
This logic will be moved to the specific `SQLDatasource` subclasses that require it.
Example (SnowflakeDatasource)

Kilo59 committed:
`quote=None`

Kilo59 committed:
distinct `_resolve_qouted_name` validator method

Kilo59 committed:
strip qoutes and re-add

Kilo59 committed:
parametrized debugging test

Kilo59 committed:
add `test_run_checkpoint`

Kilo59 committed:
add notes

Kilo59 committed:
snowflake datasource

Kilo59 committed:
add xfails for snowflake

cdkini commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -819,16 +821,36 @@ def qualified_name(self) -> str:
+            if isinstance(table_name, sqlalchemy.quoted_name):
+                return table_name
+
+            if table_name_is_qouted:
```
nit - a few instances of "quoted" are misspelled in this PR (not a big issue but would be nice to clean up 😄 )

Kilo59 replied:
Thanks, updated!

Kilo59 committed:
xfails for PG

Kilo59 committed:
`test_does_it_work`

Kilo59 committed:
mappings

Kilo59 committed:
spel better

Kilo59 committed:
update manual test script

Kilo59 commented on test_indentifiers.py:
This will be removed before merge, using it for manual testing.

Kilo59 committed:
spelling and trino

Kilo59 committed:
fix sf connection string

Kilo59 committed:
`SnowflakeTableAsset`

Kilo59 commented on great_expectations/datasource/fluent/snowflake_datasource.py:
Drop these changes as part of this PR.
Will address snowflake specific issues separately.

Kilo59 committed:
Merge branch 'develop' into b/lakitu-282/sqs_fds_qouted_identifier

Kilo59 committed:
show xpass

Kilo59 committed:
cleanup

Kilo59 committed:
drop snowflake changes

Kilo59 committed:
minimal CI tests

Kilo59 commented on tests/datasource/fluent/integration/test_sql_datasources.py:
```diff
@@ -0,0 +1,89 @@
+def postgres_ds(context: EphemeralDataContext) -> PostgresDatasource:
+    ds = context.sources.add_postgres(
+        "postgres",
+        connection_string="postgresql+psycopg2://postgres:postgres@localhost:5432/mercury",
```
Not `/mercury`

Kilo59 committed:
use container postgres

Kilo59 committed:
test_checkpoint_run

Kilo59 committed:
marks

Kilo59 committed:
Merge branch 'develop' into b/lakitu-282/sqs_fds_qouted_identifier

Kilo59 commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -819,16 +821,36 @@ def qualified_name(self) -> str:
     @pydantic.validator("table_name", pre=True, always=True)
     def _default_table_name(cls, table_name: str, values: dict, **kwargs) -> str:
         if not (validated_table_name := table_name or values.get("name")):
+            # TODO: if using `name` we should replace spaces with underscores
+            # (unless bracketed by quotes)
```
```suggestion
```

Kilo59 committed:
type ignore

Kilo59 committed:
Merge remote-tracking branch 'origin/b/lakitu-282/sqs_fds_qouted_identifier' into b/lakitu-282/sqs_fds_qouted_identifier

Kilo59 committed:
remove `table.table_name.quote. is True` check

The table_name should not be arbitrarily quoted. What's important is that it's a an `quoted_name` instance

Kilo59 commented on tests/datasource/fluent/test_sql_datasources.py:
```diff
@@ -104,7 +104,6 @@ def test_table_quoted_name_type_all_upper_case_normalizion_is_noop():
             assert str(table_asset.table_name) == table_name
             assert str(table_asset.table_name.casefold()) != table_name
             assert isinstance(table_asset.table_name, sqlalchemy.quoted_name)
-            assert table_asset.table_name.quote is True
```
This is not a useful thing to check.
What matters is that it is an instance of `sqlalchemy.quoted_name` but it doesn't (and often shouldn't) be `quote is True` (unless the original string actually was quoted).

Kilo59 committed:
fix type ignore

Kilo59 committed:
print table names

Kilo59 committed:
unused type-ignoes

Kilo59 committed:
remove comment

Kilo59 committed:
change table for CI

update with followup

Kilo59 committed:
Merge branch 'develop' into b/lakitu-282/sqs_fds_qouted_identifier

Kilo59 commented on tests/datasource/fluent/integration/test_sql_datasources.py:
```diff
@@ -0,0 +1,152 @@
+    ExpectationConfiguration,
+)
+
+PG_TABLE: Final[str] = "ge_expectations_store"
+TRINO_TABLE: Final[str] = "customer"
```
This is dependent on some existing state.
Will be issuing a follow-up PR so the test setup creates the needed tables and cleanups.

Kilo59 commented:
Also see this previous PR for column identifier changes
- https://github.com/great-expectations/great_expectations/pull/8345
